worker/src/queue.rs

```diff
@@ -179,15 +342,343 @@ impl AsRef<JsValue> for Queue {
+    pub fn with_delay_seconds(mut self, delay_seconds: u32) -> Self {
+        self.delay_seconds = Some(delay_seconds);
+        self
+    }
+
+    #[must_use]
+    /// The content type of the message.
+    /// Default is `QueueContentType::Json`.
+    pub fn with_content_type(mut self, content_type: QueueContentType) -> Self {
+        self.content_type = content_type;
+        self
+    }
+
+    /// Build the message.
+    pub fn build(self) -> SendMessage<T> {
+        SendMessage {
+            message: self.message,
+            options: Some(QueueSendOptions {
+                content_type: Some(self.content_type),
+                delay_seconds: self.delay_seconds,
+            }),
+        }
+    }
+}
+
+pub struct RawMessageBuilder {
+    message: JsValue,
+    delay_seconds: Option<u32>,
+}
+
+impl RawMessageBuilder {
+    /// Creates a new raw message builder. The message must be a `JsValue`.
+    pub fn message(message: JsValue) -> Self {
+        Self {
+            message,
+            delay_seconds: None,
+        }
+    }
+
+    #[must_use]
+    /// The number of seconds to delay a message for within the queue, before it can be delivered to a consumer
+    pub fn with_delay_seconds(mut self, delay_seconds: u32) -> Self {
+        self.delay_seconds = Some(delay_seconds);
+        self
+    }
+
+    /// Build the message with a content type.
+    pub fn build_with_content_type(self, content_type: QueueContentType) -> SendMessage<JsValue> {
+        SendMessage {
+            message: self.message,
+            options: Some(QueueSendOptions {
+                content_type: Some(content_type),
+                delay_seconds: self.delay_seconds,
+            }),
+        }
+    }
+}
+
+pub struct SendMessage<T> {
+    message: T,
+    options: Option<QueueSendOptions>,
+}
+
+impl<T: Serialize> SendMessage<T> {
+    pub fn into_raw_message(self) -> Result<SendMessage<JsValue>> {
+        Ok(SendMessage {
+            message: serde_wasm_bindgen::to_value(&self.message)?,
+            options: self.options,
+        })
+    }
+}
+
+impl<T: Serialize> From<T> for SendMessage<T> {
+    fn from(message: T) -> Self {
+        Self {
+            message,
+            options: Some(QueueSendOptions {
+                content_type: Some(QueueContentType::Json),
+                delay_seconds: None,
+            }),
+        }
+    }
+}
+
+pub struct BatchSendMessage<T> {
+    body: Vec<SendMessage<T>>,
+    options: Option<QueueSendBatchOptions>,
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+pub struct QueueSendBatchOptions {
+    delay_seconds: Option<u32>,
+}
+
+pub struct BatchMessageBuilder<T> {
+    messages: Vec<SendMessage<T>>,
+    delay_seconds: Option<u32>,
+}
+
+impl<T> BatchMessageBuilder<T> {
+    /// Creates a new batch message builder.
+    pub fn message<U: Into<SendMessage<T>>>(message: U) -> Self {
```

Comment:
I think its a little weird to have both the option to add messages in the constructor and using the builder pattern.

Reply:
👍, The `BatchMessageBuilder` now takes nothing in the `new` constructor and users can add messages using `message` or `messages` functions.

I've also updated all the other builders use `new` as a constructor so we are more consistent across them.